firedrake/slate/preconditioners.py

```diff
@@ -0,0 +1,285 @@
+        """Set up the problem context. Take the original
+        mixed problem and reformulate the problem as a
+        hybridized mixed system.
+
+        A KSP is created for the Lagrange multiplier system.
+        """
+        from ufl.algorithms.map_integrands import map_integrand_dags
+        from firedrake import (FunctionSpace, TrialFunction,
+                               TrialFunctions, TestFunction, Function,
+                               BrokenElement, MixedElement,
+                               FacetNormal, Constant, DirichletBC,
+                               Projector)
+        from firedrake.assemble import (allocate_matrix,
+                                        create_assembly_callable)
+        from firedrake.formmanipulation import ArgumentReplacer, split_form
+
+        # Extract the problem context
+        prefix = pc.getOptionsPrefix()
+        _, P = pc.getOperators()
+        context = P.getPythonContext()
+        test, trial = context.a.arguments()
+
+        V = test.function_space()
+        if V.mesh().cell_set._extruded:
+            # TODO: Merge FIAT branch to support TPC trace elements
+            raise NotImplementedError("Not implemented on extruded meshes.")
+
+        # Break the function spaces and define fully discontinuous spaces
+        broken_elements = [BrokenElement(Vi.ufl_element()) for Vi in V]
+        elem = MixedElement(broken_elements)
+        V_d = FunctionSpace(V.mesh(), elem)
+        arg_map = {test: TestFunction(V_d),
+                   trial: TrialFunction(V_d)}
+
+        # Replace the problems arguments with arguments defined
+        # on the new discontinuous spaces
+        replacer = ArgumentReplacer(arg_map)
+        new_form = map_integrand_dags(replacer, context.a)
+
+        # Create the space of approximate traces.
+        # The vector function space will have a non-empty value_shape
+        W = next(v for v in V if bool(v.ufl_element().value_shape()))
+        if W.ufl_element().family() in ["Raviart-Thomas", "RTCF"]:
+            tdegree = W.ufl_element().degree() - 1
+
+        else:
+            tdegree = W.ufl_element().degree()
+
+        # NOTE: Once extruded is ready, we will need to be aware of this
+        # and construct the appropriate trace space for the HDiv element
+        TraceSpace = FunctionSpace(V.mesh(), "HDiv Trace", tdegree)
+
+        # NOTE: For extruded, we will need to add "on_top" and "on_bottom"
+        trace_conditions = [DirichletBC(TraceSpace, Constant(0.0),
+                                        "on_boundary")]
+
+        # Set up the functions for the original, hybridized
+        # and schur complement systems
+        self.broken_solution = Function(V_d)
+        self.broken_rhs = Function(V_d)
+        self.trace_solution = Function(TraceSpace)
+        self.unbroken_solution = Function(V)
+        self.unbroken_rhs = Function(V)
+
+        # Create the symbolic Schur-reduction
+        Atilde = Tensor(new_form)
```

Comment:
Well, this is what I was wondering.  Since the broken form is only used in the forward half inside `K.T A.inv K`, and the fact that wrapping `Tensor` round it "breaks" it, might be enough?